R/ComputeFirstDifferences.R

```diff
@@ -0,0 +1,64 @@
+#' Adds first differences of a variable, robustly to gaps in time variable, as new columns in a panel dataset
+#'
+#' @param df Data frame that will be modified
+#' @param idvar Character indicating column of units.
+#' @param timevar Character indicating column of time periods.
+#' @param diffvar Character indicating column of variable whose first difference will be taken.
+#' @param timevar_holes Logical indicating whether the panel contains gaps in the time variable.
+#' @param return_df Logical indicating whether the function should return a data frame.
+#'
+#' @return The passed dataset augmented with columns that reflect the desired first differences
+#'
+#' @seealso [data.table::shift()]
+#'
+#' @examples
+#' ComputeFirstDifferences(
+#'     example_data,
+#'     idvar = "id",
+#'     timevar = "t",
+#'     diffvar = "z"
+#' )
+#'
+#' @importFrom data.table setDT setorderv setnames shift := CJ .SD
+#'
+#' @keywords internal
+#' @noRd
+
+ComputeFirstDifferences <- function(df, idvar, timevar, diffvar,
+                                    timevar_holes = FALSE, return_df = TRUE) {
+    if (! is.data.frame(df)) {stop("df should be a data frame.")}
+    if ((! is.null(idvar)) & (! is.character(idvar))) {stop("idvar should be a character.")}
+    if (! is.character(timevar)) {stop("timevar should be a character.")}
+    if (! is.character(diffvar)) {stop("diffvar should be a character.")}
+    if (! is.logical(timevar_holes)) {stop("timevar_holes should be a logical.")}
+
+    data.table::setDT(df)
+    data.table::setorderv(df, cols = c(idvar, timevar))
+
+    if (!timevar_holes) {
+        df[, paste0(diffvar, "_fd") := get(diffvar) - shift((get(diffvar))),
+           by = idvar]
+    } else {
+        ## Create dataset with all combinations to compute first differences
+        all_combinations <- CJ(unique(df[[idvar]]), unique(df[[timevar]]))
```

Comment:
Thanks @nateschor! 

Let me start by clarifying that your example is a bit off. You should use `CJ(unique(test_i), unique(test_t))`, in which case you code would result in
```Routput
   V1 V2
1:  1  1
2:  1  2
3:  1  3
4:  1  5
5:  2  1
6:  2  2
7:  2  3
8:  2  5
```

Regardless of this issue, you raise a very reasonable concern. Two comments:

- The time period `4` is missing! This would not happen if at least one `i` has all values of `t`, which I expect to be the norm. But examples like yours could arise, so we need to fix this. To get the `4` in your example we need to modify L43 to:
  ```R
  all_combinations <- CJ(unique(df[[idvar]]), min(df[[timevar]]):max(df[[timevar]]))
  ```
  This only works because timevar is forced to be an integer, otherwise things would get a bit more complicated. I implemented this in https://github.com/JMSLab/eventstudyr/pull/38/commits/fad3f739dcaf3b2f9bb263b415ef2659640fd85c.
- Even if the 4 was there, unit 2 will have unnecessary rows that will always be NA. This is correct, but it's not a problem. Those unwanted rows will be dropped when we bring all combinations back to the original dataset below:
  https://github.com/JMSLab/eventstudyr/blob/fad3f739dcaf3b2f9bb263b415ef2659640fd85c/R/ComputeShifts.R#L92-L93
  The `all.x = TRUE` keeps only rows from `dt` in the output.

  I don't think that changing the code by searching the maximum and minimum `t` for each `i` would be an improvement here. This would (1) complicate the code, (2) probably yield small, if any, speed gains (you don't carry over unnecessary rows, but you need more computations to find maximum and minimum for each `i`). The one gain would be using less RAM, which would only become relevant for big datasets.

Below is your example after the fixes. Curious to hear your thoughts @nateschor!

<details>
  <summary>Example after fixes</summary>

```R
test_i <- c(1, 2)
test_t <- c(1:3, 5, 1:3)
CJ(unique(test_i), min(test_t):max(test_t))

# output
    V1 V2
 1:  1  1
 2:  1  2
 3:  1  3
 4:  1  4
 5:  1  5
 6:  2  1
 7:  2  2
 8:  2  3
 9:  2  4
10:  2  5
```

</details>